### PR TITLE
Fixed udi's rendered in rte content

### PR DESCRIPTION
**Pull request by dawoe**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have linked this PR to an issue on the tracker at http://issues.umbraco.org/issue/U4-11140

### Description

Fixed this in the TemplateUtilities.ParseInternalLinks method. First attempt was to do this in the PropertyValueConverter for the RTE, but then issue still rises in Rich text in the grid. 

New code is also covered by new test cases

### CODE REVIEW

**Comment by emmaburstow** [comment]
Hi @dawoe 

Thanks for the PR! We'll have a look at it and let you know if we need any changes. Have a good one!

Emma, [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-pr-team/)

**Comment by dawoe** [comment]
Okay..didn't know that one. But for me it seems like a security issue. You make it possible to identify a site running Umbraco and are actually leaking data.

**Comment by nul800sebastiaan** [comment]
@Jeavon Do the udi's need to be rendered to the frontend though?
Maybe we can then compromise and just leave the GUID in place but not the schema?

**Comment by Jeavon** [comment]
No. We need them to be able to find the media item that has been embedded into the RTE. We do then have a option to not render them to the eventual output of the website. 

https://github.com/Jeavon/Slimsy/blob/develop/Slimsy/Slimsy.cs#L227
https://github.com/Jeavon/Slimsy/blob/develop/Slimsy/Slimsy.cs#L251

**Comment by dawoe** [comment]
What if we made the stripping of the udi it configurable ? By the default we can set it to on. You can instruct slimsy users to disable it and do the stripping your self ?

Dave

**Comment by Jeavon** [comment]
@dawoe I have been thinking long and hard about this and also have discussed it with Seb. It's certainly correct that Umbraco shouldn't render this data to the front end. However I think the logic should be moved from the ParseInternalLinks method to the ConvertDataToSource method of RteMacroRenderingValueConverter,cs as the HtmlAgilityPack is already being used there to remove "rel" attributes for upgraded sites (see http://issues.umbraco.org/issue/U4-5480) and it's best to avoid HtmlAgility being loaded more than once. 

This is going to cause me a headache for packages that rely on this such as Slimsy and AzureCDNToolkit as we don't have "stacked" value converters so if you were using both packages it's not going to work but I'll have to work around it somehow.

I will want to test this out before it's released to ensure that it will be possible, I'm also going to have to have a version of Slimsy ready to release as upgraded sites will stop working!

Jeavon

**Comment by dawoe** [comment]
Hi @Jeavon 

My first attempt was actually doing it int PropertyValueConverter...but that did not solve the problem for grid RTE block. That's why added it to the ParseInternalLinks method

Dave

**Comment by Jeavon** [comment]
@dawoe that's a real pain! Perhaps a new method is needed in TemplateUtilities named CleanMediaAttributes or similar so that there is only one usage of HtmlAgility. This would then be called from the value converter and also added to the RET.cshtml Grid partial view?

**Comment by dawoe** [comment]
Hi @jeavon and @nul800sebastiaan 

I'm back from holiday and want to pick this one up again. I'm thinking about doing it this way.

- Add a method to TemplateUtilities that strips udi data attributes with 2 overloads accepting a htmlagility htmldoc or a string. That way it can be called from the PropertyValueConverter for the RTE as well from the ParseInternalLinks method
- Add a extra optional parameter to the ParseInternalLinks method to indicate udi data attributes should be stripped or not
- Add a config setting that will enable/disable stripping of data-udi attributes. The default would be true

This way slimsy could disable the stripping of the data-attributes and use the new method on the templateutilities to strip them when doing there stuff

I think this way we don't introduce any breaking changes in grid layouts or other code that is calling ParseInternalLinks

What do you guys think

Dave

**Comment by nul800sebastiaan** [comment]
@dawoe 

Do we need two overloads in `TemplateUtilities`? Sorry, I haven't looked into the code recently, we're already stripping stuff with HTML Agility somewhere, can't we just strip the UDIs there as well? Maybe I'm missing something.

We want to avoid adding even more configuration options to Umbraco, especially ones that are not that useful for most people. 
This would still require a new release of Slimsy to be sure that the stripping is being disabled, so it will be just as breaking for now.

**Comment by dawoe** [comment]
@nul800sebastiaan 

We need the 2 overloads...one will be called from the PVC where we already have html agility htmldoc instance.

But the ParseInternalLinks method is also called in other parts (eg grid rendering) and there we have string, that first needs to be converted to a html doc.

I wanted to add the setting so Slimsy doesn't need to create a custom PVC for the RTE

Dave

**Comment by nul800sebastiaan** [comment]
Ah cool, overloads are fine then! I'll let @Jeavon comment on the need for a PVC, what do you think.. another config value or not?

**Comment by dawoe** [comment]
Hi @Jeavon and @nul800sebastiaan 

I refactored the code so there is only one pass through html agility pack from the RTE PVC

I did not add the setting yet for disabling the stripping

Dave

File: src/Umbraco.Web/Templates/TemplateUtilities.cs [hunk 1]
```diff
@@ -74,16 +82,22 @@ public static string ParseInternalLinks(string text, UrlProvider urlProvider)
...
 
         /// <summary>
         /// Parses the string looking for the {localLink} syntax and updates them to their correct links.
         /// </summary>
         /// <param name="text"></param>
+        /// <param name="stripUdiDataAttributes"></param>
         /// <returns></returns>
         [Obsolete("Use the overload specifying all dependencies instead")]
-        public static string ParseInternalLinks(string text)
+        public static string ParseInternalLinks(string text,bool stripUdiDataAttributes = true)
```
Comment (Jeavon) [inline src/Umbraco.Web/Templates/TemplateUtilities.cs]: Adding new parameters to a obsolete method isn't ideal. Is it necessary?

Comment (dawoe) [reply src/Umbraco.Web/Templates/TemplateUtilities.cs]: Yep without doing a lot of refactoring in the PVC this is needed. This is the method called by several PVC

Dave

**Comment by dawoe** [comment]
I see the appveyor build failed on a unit test. But that one is green on my machine

**Comment by nul800sebastiaan** [comment]
@dawoe That is an unrelated test, that is failing because I merged something into `dev-v7`, I'll fix it up later today.

Anyway, as @Jeavon pointed out, the method change is not optimal and would be fine, if it is a small change to support this PR. However, it is a change to the method signature of a public class. This is a breaking change, even if the new parameter is optional. This will require everyone using the method to have to rebuild their custom code and we want to avoid that.

I think at this point it would be best to just introduce that setting and I'll accept that it has to be that way (otherwise we'll have to introduce 2 new overloads of already obsolete methods which... is very messy).

**Comment by dawoe** [comment]
Hi @nul800sebastiaan and @Jeavon 

I removed the changes in the method signatures for ParseInternalLinks back to the previous version.
Stripping the data-udi attributes is now done using a regex instead of HtmlAgility. Should have thought of this in the first place 

I also added the new setting for disabling this.

Add this to the content section of umbracoSettings.config to disable it
 <EnableStrippingUdiDataAttributes>false</EnableStrippingUdiDataAttributes>

Dave

**Comment by nul800sebastiaan** [comment]
Unit test is fixed now, I'll have a look at the current state soon! Probably not today though :-)

**Comment by dawoe** [comment]
Any update on this one ? Anything needed from me ? I have busy month coming up :-)

**Comment by nul800sebastiaan** [comment]
For @Jeavon - we'll update this to `StripUdiAttributes` and put an explanation in the config file.

**Comment by nul800sebastiaan** [comment]
All updated, squashed and merged! 👍

**Comment by readingdancer** [comment]
Hi @nul800sebastiaan 

I've just come across this in a site I am currently building that is due to go live using Umbraco Cloud, it would be good to know why this has been reverted and if there is any way to remove the unwanted data-udi attributes? I agree that this is a bit of a security risk for our clients which is annoying if there is now no fix being released.

Cheers, Chris

**Comment by nul800sebastiaan** [comment]
@readingdancer If it really a big concern you can always strip them out yourself before rendering the property. We realized that not only Slimsy relied on this attribute being here but other things too and it was just too soon to prevent this attribute from being rendered.

**Comment by dawoe** [comment]
Hi @readingdancer 

I have custom property value converter for the RTE that does this. It only does it for links because we don't allow images. But the code can easily be adapted 

```
using System.Web;
    using System.Web.Mvc;

    using HtmlAgilityPack;

    using Umbraco.Core.Models.PublishedContent;
    using Umbraco.Core.PropertyEditors;

    /// <summary>
    /// The rte value converter.
    /// </summary>
    [PropertyValueType(typeof(IHtmlString))]
    [PropertyValueCache(PropertyCacheValue.All, PropertyCacheLevel.Request)]
    internal class RteValueConverter : Umbraco.Web.PropertyEditors.ValueConverters.RteMacroRenderingValueConverter
    {
        /// <summary>Converts a property Data value to a Source value.</summary>
        /// <param name="propertyType">The property type.</param>
        /// <param name="source">The data value.</param>
        /// <param name="preview">A value indicating whether conversion should take place in preview mode.</param>
        /// <returns>The result of the conversion.</returns>
        /// <remarks>
        /// <para>The converter should know how to convert a <c>null</c> raw value, meaning that no
        /// value has been assigned to the property. The source value can be <c>null</c>.</para>
        /// <para>With the XML cache, raw values come from the XML cache and therefore are strings.</para>
        /// <para>With objects caches, raw values would come from the database and therefore be either
        /// ints, DateTimes, or strings.</para>
        /// <para>The converter should be prepared to handle both situations.</para>
        /// <para>When raw values are strings, the converter must handle empty strings, whitespace
        /// strings, and xml-whitespace strings appropriately, ie it should know whether to preserve
        /// whitespaces.</para>
        /// </remarks>
        public override object ConvertDataToSource(PublishedPropertyType propertyType, object source, bool preview)
        {
            if (source == null)
            {
                return null;
            }

            var sourceString = base.ConvertDataToSource(propertyType, source, preview).ToString();

            if (!string.IsNullOrEmpty(sourceString))
            {
                var content = new HtmlDocument();
                content.LoadHtml(sourceString);

                var modified = false;

                var anchors = content.DocumentNode.SelectNodes("//a");

                if (anchors != null)
                {
                    foreach (var link in anchors)
                    {
                        var titleAttribute = link.Attributes["title"];                       

                        if (string.IsNullOrEmpty(titleAttribute?.Value))
                        {
                            link.SetAttributeValue("title", link.InnerText);
                            modified = true;
                        }

                        var dataUdi = link.Attributes["data-udi"];

                        if (dataUdi != null)
                        {
                            link.Attributes.Remove("data-udi");
                            modified = true;
                        }
                    }

                    if (modified)
                    {
                        sourceString = content.DocumentNode.OuterHtml;
                    }
                   
                }

            }

            return sourceString;
        }
    }
}
```

Dave

**Comment by fbenyah** [comment]
Hello,

So is the configurable option still available in v8 to keep the parsed value the udi and not the actual path?

Hoping to get something like <a href="umb://document/eecf6327430a43cc9344a737f27c779f">My Link</a>

Thanks

**Comment by dawoe** [comment]
Hi @fbenyah,

The PR got reverted. So I don't think it has ever been merged for V8.

Dave

**Comment by fbenyah** [comment]
Hi @dawoe 

Thanks for the response. Do you know if there is a work around at the moment?

**Pull request merged**